In [1]:
# 1. import of numpy and pandas:
import numpy as np
import pandas as pd

In [2]:
df = pd.read_csv('../data/nic-house-prices.csv')

In [3]:
df.shape

(2430, 82)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2430 entries, 0 to 2429
Data columns (total 82 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Order            2430 non-null   int64  
 1   PID              2430 non-null   int64  
 2   MS SubClass      2430 non-null   int64  
 3   MS Zoning        2430 non-null   object 
 4   Lot Frontage     2010 non-null   float64
 5   Lot Area         2430 non-null   int64  
 6   Street           2430 non-null   object 
 7   Alley            163 non-null    object 
 8   Lot Shape        2430 non-null   object 
 9   Land Contour     2430 non-null   object 
 10  Utilities        2430 non-null   object 
 11  Lot Config       2430 non-null   object 
 12  Land Slope       2430 non-null   object 
 13  Neighborhood     2430 non-null   object 
 14  Condition 1      2430 non-null   object 
 15  Condition 2      2430 non-null   object 
 16  Bldg Type        2430 non-null   object 
 17  House Style   

In [11]:
df.head()

Order        PID  MS SubClass MS Zoning  Lot Frontage  Lot Area Street  \
0    484  528275070           60        RL           NaN      8795   Pave   
1   2586  535305120           20        RL          75.0     10170   Pave   
2   2289  923228250          160        RM          21.0      2001   Pave   
3    142  535152150           20        RL          70.0     10552   Pave   
4   2042  903475060          190        RM          60.0     10120   Pave   

  Alley Lot Shape Land Contour  ... Screen Porch Pool Area Pool QC  Fence  \
0   NaN       IR1          Lvl  ...            0         0     NaN    NaN   
1   NaN       Reg          Lvl  ...            0         0     NaN    NaN   
2   NaN       Reg          Lvl  ...            0         0     NaN    NaN   
3   NaN       IR1          Lvl  ...            0         0     NaN    NaN   
4   NaN       IR1          Bnk  ...            0         0     NaN  MnPrv   

  Misc Feature Misc Val Mo Sold Yr Sold  Sale Type  Sale Condition  
0          NaN        0       4    2009        WD           Normal  
1          NaN        0       6    2006        WD           Normal  
2          NaN        0       1    2007        WD           Normal  
3          NaN        0       4    2010        WD           Normal  
4          NaN        0       1    2007        WD           Normal  

[5 rows x 81 columns]

In [6]:
df.tail()

Order        PID  MS SubClass MS Zoning  Lot Frontage  Lot Area Street  \
2425    764  904301070           90        RL          60.0     10800   Pave   
2426    836  907126010           20        RL          82.0     12464   Pave   
2427   1654  527353080           60        RL          80.0     11584   Pave   
2428   2608  535402070           80        RL          85.0     13400   Pave   
2429   2733  905402110           20        RL          75.0      9375   Pave   

     Alley Lot Shape Land Contour  ... Pool Area Pool QC  Fence Misc Feature  \
2425   NaN       Reg          Lvl  ...         0     NaN    NaN          NaN   
2426   NaN       IR2          Low  ...         0     NaN  GdPrv          NaN   
2427   NaN       Reg          Lvl  ...         0     NaN    NaN          NaN   
2428   NaN       Reg          Lvl  ...         0     NaN   GdWo          NaN   
2429   NaN       Reg          Lvl  ...         0     NaN    NaN          NaN   

     Misc Val Mo Sold Yr Sold Sale Type  Sale Condition  SalePrice  
2425        0       3    2009       WD           Alloca     179000  
2426        0      11    2009       WD           Normal     152000  
2427        0      11    2007       WD           Normal     197000  
2428        0       6    2006       WD           Normal     159950  
2429        0       6    2006       WD           Normal     146000  

[5 rows x 82 columns]

In [11]:
df.dtypes

Order               int64
PID                 int64
MS SubClass         int64
MS Zoning          object
Lot Frontage      float64
                   ...   
Mo Sold             int64
Yr Sold             int64
Sale Type          object
Sale Condition     object
SalePrice           int64
Length: 82, dtype: object

## Explore the data and examples of special types of variables
nominal variables
* Street
* Sale Type

ordinal variables
* Order
* Year Sold

discrete variables
* Mo Sold

continuous variables
* Lot Frontage?

In [19]:
# ChatGPT's solution:
# nominal variables
nominal_cols = [col for col in df.columns if df[col].nunique() < 10]
print('Nominal variables: ', nominal_cols)

# ordinal variables
ordinal_cols = [col for col in df.columns if df[col].dtype == 'object' and df[col].nunique() > 10]
print('Ordinal variables: ', ordinal_cols)

# discrete variables
discrete_cols = [col for col in df.columns if df[col].dtype == 'int64']
print('Discrete variables: ', discrete_cols)

# continuous variables
continuous_cols = [col for col in df.columns if df[col].dtype == 'float64']
print('Continuous variables: ', continuous_cols)

Nominal variables:  ['MS Zoning', 'Street', 'Alley', 'Lot Shape', 'Land Contour', 'Utilities', 'Lot Config', 'Land Slope', 'Condition 1', 'Condition 2', 'Bldg Type', 'House Style', 'Overall Cond', 'Roof Style', 'Roof Matl', 'Mas Vnr Type', 'Exter Qual', 'Exter Cond', 'Foundation', 'Bsmt Qual', 'Bsmt Cond', 'Bsmt Exposure', 'BsmtFin Type 1', 'BsmtFin Type 2', 'Heating', 'Heating QC', 'Central Air', 'Electrical', 'Bsmt Full Bath', 'Bsmt Half Bath', 'Full Bath', 'Half Bath', 'Bedroom AbvGr', 'Kitchen AbvGr', 'Kitchen Qual', 'Functional', 'Fireplaces', 'Fireplace Qu', 'Garage Type', 'Garage Finish', 'Garage Cars', 'Garage Qual', 'Garage Cond', 'Paved Drive', 'Pool QC', 'Fence', 'Misc Feature', 'Yr Sold', 'Sale Condition']
Ordinal variables:  ['Neighborhood', 'Exterior 1st', 'Exterior 2nd']
Discrete variables:  ['Order', 'PID', 'MS SubClass', 'Lot Area', 'Overall Qual', 'Overall Cond', 'Year Built', 'Year Remod/Add', '1st Flr SF', '2nd Flr SF', 'Low Qual Fin SF', 'Gr Liv Area', 'Full Bath',

## Create a boolean mask to identify houses in the dataset where the 'Sale Condition' is set to 'Abnorml'

In [22]:
df_abnorml = df[df['Sale Condition']=='Abnorml']
df_abnorml

Order        PID  MS SubClass MS Zoning  Lot Frontage  Lot Area Street  \
65      907  909129100           50        RH          55.0      4500   Pave   
72     2598  535376100           30        RL          60.0      5400   Pave   
84     2591  535325460           30        RL         118.0     21299   Pave   
89     1573  916325080           20        RL           NaN     15498   Pave   
95     1372  904100020           20        RL          82.0      9488   Pave   
...     ...        ...          ...       ...           ...       ...    ...   
2386    152  535304060          190        RL          70.0     13587   Pave   
2390    545  531382070           20        RL          88.0     10994   Pave   
2413   2223  909428120           20        RL           NaN     21000   Pave   
2419   2648  902125020           70        RM          50.0      9000   Pave   
2422    278  908102300           50        RL          89.0     11792   Pave   

     Alley Lot Shape Land Contour  ... Pool Area Pool QC  Fence Misc Feature  \
65    Pave       IR2          Bnk  ...         0     NaN    NaN          NaN   
72     NaN       Reg          Lvl  ...         0     NaN    NaN          NaN   
84     NaN       Reg          Lvl  ...         0     NaN    NaN          NaN   
89     NaN       IR1          Lvl  ...         0     NaN    NaN          NaN   
95     NaN       Reg          Lvl  ...         0     NaN    NaN          NaN   
...    ...       ...          ...  ...       ...     ...    ...          ...   
2386   NaN       Reg          Lvl  ...         0     NaN    NaN          NaN   
2390   NaN       IR1          Lvl  ...         0     NaN    NaN          NaN   
2413   NaN       Reg          Bnk  ...         0     NaN  MnPrv          NaN   
2419  Grvl       Reg          Lvl  ...         0     NaN  MnPrv          NaN   
2422   NaN       Reg          Lvl  ...         0     NaN    NaN          NaN   

     Misc Val Mo Sold Yr Sold Sale Type  Sale Condition  SalePrice  
65          0       7    2009       WD          Abnorml     159434  
72          0      12    2006       WD          Abnorml     108480  
84          0      12    2006       COD         Abnorml     167000  
89          0       5    2008       COD         Abnorml     287000  
95          0       9    2008       WD          Abnorml     124000  
...       ...     ...     ...       ...             ...        ...  
2386        0       4    2010       WD          Abnorml     119000  
2390        0       9    2009       COD         Abnorml     257000  
2413        0      10    2007       COD         Abnorml     217000  
2419        0       6    2006       WD          Abnorml     124000  
2422        0       6    2010       WD          Abnorml      64000  

[154 rows x 82 columns]

In [25]:
df_abnorml['Lot Frontage'].info()

<class 'pandas.core.series.Series'>
Int64Index: 154 entries, 65 to 2422
Series name: Lot Frontage
Non-Null Count  Dtype  
--------------  -----  
132 non-null    float64
dtypes: float64(1)
memory usage: 2.4 KB


In [26]:
df_abnorml['Lot Frontage'].describe()

count    132.000000
mean      70.818182
std       24.802985
min       21.000000
25%       56.750000
50%       68.000000
75%       82.000000
max      174.000000
Name: Lot Frontage, dtype: float64

132 / 154 entries are non-null.

I would rather replace the missing values with the median because it seems to be ordinal values (can't operate maths on it).

In [27]:
COLUMN_NAME = 'Lot Frontage'
df[COLUMN_NAME].fillna(df[COLUMN_NAME].median(), inplace=True)

df[COLUMN_NAME].info()

<class 'pandas.core.series.Series'>
RangeIndex: 2430 entries, 0 to 2429
Series name: Lot Frontage
Non-Null Count  Dtype  
--------------  -----  
2430 non-null   float64
dtypes: float64(1)
memory usage: 19.1 KB


In [ ]:
# Outliers
def z_socre(x):
    mean_x = np.mean(x)

    std_x = np.std(x)
    return ...

In [ ]:
df